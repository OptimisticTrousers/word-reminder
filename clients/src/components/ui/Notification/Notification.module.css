.notification {
  border-radius: 6px;
  padding: 1rem;
}

.notification--success {
  --text-color: var(--success-text-color);
  --icon-color: var(--success-icon-color);
  --button-bkg-color: var(--success-button-bkg-color);
  --secondary-button-bkg-color: var(--secondary-success-button-bkg-color);
  background-color: var(--success-bkg-color);
}

.notification--error {
  --text-color: var(--alert-text-color);
  --icon-color: var(--alert-icon-color);
  --button-bkg-color: var(--alert-button-bkg-color);
  --secondary-button-bkg-color: var(--secondary-alert-button-bkg-color);
  background-color: var(--alert-bkg-color);
}

.notification__flex {
  display: flex;
  align-items: center;
}

.notification__item {
  flex-shrink: 0;
}

.notification__box {
  margin-left: 0.75rem;
}

.notification__message {
  color: var(--text-color);
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.25rem;
  display: block;
}

.notification__controls {
  margin-left: auto;
  padding-left: 0.75rem;
}

.notification__control {
  margin-top: -0.375rem;
  margin-right: -0.375rem;
  margin-bottom: -0.375rem;
  margin-left: -0.375rem;
}

.notification__button {
  color: var(--text-color);
  padding: 0.375rem;
  background-color: var(--button-bkg-color);
  border-radius: 6px;
  display: inline-flex;
}

.notification__button:hover {
  background-color: var(--secondary-button-bkg-color);
}

.notification__text {
  position: absolute;
  width: 1px;
  padding: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border-width: 0;
}

.notification__icon {
  width: 1.25rem;
  height: 1.25rem;
}

.notification__icon--circle {
  color: var(--icon-color);
}
