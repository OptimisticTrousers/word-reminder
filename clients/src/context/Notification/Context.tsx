import { createContext } from "react";

import { NotificationContext as Context } from "./Provider";

export enum NOTIFICATION_ACTIONS {
  SUCCESS = "success",
  ERROR = "error",
}

export const NotificationContext = createContext<Context>({
  showNotification: function () {},
  dismissNotification: function () {},
});
