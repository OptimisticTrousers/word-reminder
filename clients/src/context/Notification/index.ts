export { NOTIFICATION_ACTIONS, NotificationContext } from "./Context";
export { NotificationProvider } from "./Provider";
