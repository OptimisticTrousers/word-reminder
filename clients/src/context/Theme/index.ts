export { ThemeContext } from "./Context";
export { ThemeProvider } from "./Provider";
