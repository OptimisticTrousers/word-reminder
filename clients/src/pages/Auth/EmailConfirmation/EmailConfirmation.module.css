.email-confirmation {
  padding: 1rem;
}

.email-confirmation__container {
  margin-bottom: 1rem;
}

.email-confirmation__heading {
  margin-bottom: 0.75rem;
}

.email-confirmation__image {
  width: 250px;
  height: 250px;
}

.email-confirmation__bold {
  font-weight: 600;
}

.email-confirmation__button {
  margin-top: 1rem;
}
