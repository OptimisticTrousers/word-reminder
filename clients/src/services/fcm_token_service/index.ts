export { fcmTokenService } from "./fcm_token_service";
