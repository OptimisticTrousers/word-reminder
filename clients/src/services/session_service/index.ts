export { sessionService } from "./session_service";
