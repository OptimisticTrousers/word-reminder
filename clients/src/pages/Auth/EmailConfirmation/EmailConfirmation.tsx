import { Subject, Template, User } from "common";
import CSSModules from "react-css-modules";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { emailService } from "../../../services/email_service";
import styles from "./EmailConfirmation.module.css";
import { Loading } from "../../../components/ui/Loading";
import { Error500 } from "../../Error500";
import { sessionService } from "../../../services/session_service";
import { extension } from "../../../utils/extension";
import { useNotificationError } from "../../../hooks/useNotificationError";
import { useNavigate } from "react-router-dom";

interface Props {
  user: User;
}

export const EmailConfirmation = CSSModules(
  function ({ user }: Props) {
    const userId = String(user.id);
    const { showNotificationError } = useNotificationError();
    const queryClient = useQueryClient();
    const navigate = useNavigate();
    const { isLoading, failureReason } = useQuery({
      queryKey: ["emails"],
      queryFn: () => {
        return emailService.sendEmail({
          userId,
          body: {
            email: user.email,
            subject: Subject.CONFIRM_ACCOUNT,
            template: Template.CONFIRM_ACCOUNT,
          },
        });
      },
    });
    const { isPending, mutate } = useMutation({
      mutationFn: sessionService.logoutUser,
      onSettled: async () => {
        await extension.storage.sync.remove("userId");
        await navigate("/login");
      },
      onError: showNotificationError,
    });

    async function handleLogout() {
      mutate();
      queryClient.clear();
    }

    if (failureReason) {
      return <Error500 message={failureReason.message} />;
    }

    if (isLoading) {
      return <Loading />;
    }

    const disabled = isPending;

    return (
      <div styleName="email-confirmation">
        <div styleName="email-confirmation__container">
          <img
            styleName="email-confirmation__image"
            src="/images/word-reminder.png"
            alt=""
          />
        </div>
        <h2 styleName="email-confirmation__heading">Check your email</h2>
        <p styleName="email-confirmation__message">
          Follow the link in the email sent to
          <span styleName="email-confirmation__bold"> {user.email} </span>
          to create, verify, and use your account.
        </p>
        <button
          styleName="email-confirmation__button"
          onClick={handleLogout}
          disabled={disabled}
        >
          Cancel
        </button>
      </div>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
