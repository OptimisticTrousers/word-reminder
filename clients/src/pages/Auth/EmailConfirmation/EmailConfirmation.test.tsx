import { render, screen, waitFor } from "@testing-library/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

import { emailService } from "../../../services/email_service";
import { EmailConfirmation } from "./EmailConfirmation";
import { Subject, Template } from "common";
import { NotificationProvider } from "../../../context/Notification";
import { createRoutesStub, Outlet } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { sessionService } from "../../../services/session_service";

vi.mock("../../../components/ui/Loading/Loading");

vi.mock("../../Error500/Error500");

describe("EmailConfirmation component", () => {
  const info = {
    message:
      "Message sent: <d786aa62-4e0a-070a-47ed-0b0666549519@ethereal.email",
  };

  const user = {
    id: 1,
    email: "bob@gmail.com",
    confirmed: false,
    created_at: new Date(),
    updated_at: new Date(),
  };

  const status = 200;

  function setup(queryClient: QueryClient) {
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return (
            <NotificationProvider>
              <QueryClientProvider client={queryClient}>
                <Outlet />
              </QueryClientProvider>
            </NotificationProvider>
          );
        },
        children: [
          {
            path: "/email-confirmation",
            Component: function () {
              return <EmailConfirmation user={user} />;
            },
          },
          {
            path: "/login",
            Component: function () {
              return <div data-testid="login"></div>;
            },
          },
        ],
      },
    ]);

    return {
      user: userEvent.setup(),
      ...render(<Stub initialEntries={["/email-confirmation"]} />),
    };
  }

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("calls the functions to send confirmation email", async () => {
    const mockSendEmail = vi
      .spyOn(emailService, "sendEmail")
      .mockImplementation(async () => {
        return { json: { info }, status };
      });
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });

    const { asFragment } = setup(queryClient);

    const heading = await screen.findByRole("heading", {
      name: "Check your email",
    });
    const message = screen.getByText((_, element) => {
      return (
        element?.textContent ===
        `Follow the link in the email sent to ${user.email} to create, verify, and use your account.`
      );
    });
    expect(heading).toBeInTheDocument();
    expect(message).toBeInTheDocument();
    expect(mockSendEmail).toHaveBeenCalledTimes(1);
    expect(mockSendEmail).toHaveBeenCalledWith({
      userId: String(user.id),
      body: {
        email: user.email,
        subject: Subject.CONFIRM_ACCOUNT,
        template: Template.CONFIRM_ACCOUNT,
      },
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows that it is sending email when loading", async () => {
    const delay = 50;
    const mockSendEmail = vi
      .spyOn(emailService, "sendEmail")
      .mockImplementation(async () => {
        return new Promise((resolve) => {
          setTimeout(() => {
            resolve({ json: { info }, status });
          }, delay);
        });
      });
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });

    setup(queryClient);

    const loading = screen.getByTestId("loading");
    expect(loading).toBeInTheDocument();
    expect(mockSendEmail).toHaveBeenCalledTimes(1);
    expect(mockSendEmail).toHaveBeenCalledWith({
      userId: String(user.id),
      body: {
        email: user.email,
        subject: Subject.CONFIRM_ACCOUNT,
        template: Template.CONFIRM_ACCOUNT,
      },
    });
  });

  it("shows that it is sending email when there is an error", async () => {
    const message = "Bad Request.";
    const status = 400;
    const mockSendEmail = vi
      .spyOn(emailService, "sendEmail")
      .mockImplementation(async () => {
        return Promise.reject({ json: { message }, status });
      });
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });

    setup(queryClient);

    const error = await screen.findByTestId("error-500");
    expect(error).toBeInTheDocument();
    expect(mockSendEmail).toHaveBeenCalledTimes(1);
    expect(mockSendEmail).toHaveBeenCalledWith({
      userId: String(user.id),
      body: {
        email: user.email,
        subject: Subject.CONFIRM_ACCOUNT,
        template: Template.CONFIRM_ACCOUNT,
      },
    });
  });

  describe("when cancelling", () => {
    it("logs out the user when the 'Cancel' button is clicked", async () => {
      vi.spyOn(emailService, "sendEmail").mockImplementation(async () => {
        return { json: { info }, status };
      });
      const queryClient = new QueryClient({
        defaultOptions: { queries: { retry: false } },
      });
      const mockClear = vi.spyOn(queryClient, "clear");
      const mockLogout = vi
        .spyOn(sessionService, "logoutUser")
        .mockImplementation(async () => {
          return { json: { user: { id: "1" } }, status: 200 };
        });
      const mockStorageRemove = vi.spyOn(window.chrome.storage.sync, "remove");
      const { user } = setup(queryClient);

      const cancelButton = await screen.findByRole("button", {
        name: "Cancel",
      });
      await user.click(cancelButton);

      const login = await screen.findByTestId("login");
      expect(login).toBeInTheDocument();
      expect(mockStorageRemove).toHaveBeenCalledTimes(1);
      expect(mockStorageRemove).toHaveBeenCalledWith("userId");
      expect(mockLogout).toHaveBeenCalledTimes(1);
      expect(mockLogout).toHaveBeenCalledWith(undefined);
      expect(mockClear).toHaveBeenCalledTimes(1);
      expect(mockClear).toHaveBeenCalledWith();
    });

    it("disables 'Cancel' button when mutation is pending", async () => {
      vi.spyOn(emailService, "sendEmail").mockImplementation(async () => {
        return { json: { info }, status };
      });
      const delay = 500;
      const mockLogout = vi
        .spyOn(sessionService, "logoutUser")
        .mockImplementation(async () => {
          return new Promise((resolve) => {
            setTimeout(() => {
              resolve({ json: { user: { id: "1" } }, status: 200 });
            }, delay);
          });
        });
      const queryClient = new QueryClient({
        defaultOptions: { queries: { retry: false } },
      });
      const { user } = setup(queryClient);

      await waitFor(async () => {
        const cancelButton = screen.getByRole("button", {
          name: "Cancel",
        });
        await user.click(cancelButton);

        expect(cancelButton).toBeDisabled();
        expect(mockLogout).toHaveBeenCalledTimes(1);
        expect(mockLogout).toHaveBeenCalledWith(undefined);
      });
    });

    it("when there is an error, it shows a notification error when attempting to cancel", async () => {
      vi.spyOn(emailService, "sendEmail").mockImplementation(async () => {
        return { json: { info }, status };
      });
      const message = "Bad Request.";
      const status = 400;
      const mockLogout = vi
        .spyOn(sessionService, "logoutUser")
        .mockImplementation(async () => {
          return Promise.reject({ json: { message }, status });
        });
      const queryClient = new QueryClient({
        defaultOptions: { queries: { retry: false } },
      });
      const { user } = setup(queryClient);

      const cancelButton = await screen.findByRole("button", {
        name: "Cancel",
      });
      await user.click(cancelButton);

      const notification = await screen.findByRole("dialog", { name: message });
      expect(notification).toBeInTheDocument();
      expect(mockLogout).toHaveBeenCalledTimes(1);
      expect(mockLogout).toHaveBeenCalledWith(undefined);
    });
  });
});
