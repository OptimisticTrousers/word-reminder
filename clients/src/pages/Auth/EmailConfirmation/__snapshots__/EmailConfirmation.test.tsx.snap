// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmailConfirmation component > calls the functions to send confirmation email 1`] = `
<DocumentFragment>
  <div
    class="_email-confirmation_9d1292"
  >
    <div
      class="_email-confirmation__container_9d1292"
    >
      <img
        alt=""
        class="_email-confirmation__image_9d1292"
        src="/images/word-reminder.png"
      />
    </div>
    <h2
      class="_email-confirmation__heading_9d1292"
    >
      Check your email
    </h2>
    <p
      class="_email-confirmation__message_9d1292"
    >
      Follow the link in the email sent to
      <span
        class="_email-confirmation__bold_9d1292"
      >
         bob@gmail.com 
      </span>
      to create, verify, and use your account.
    </p>
    <button
      class="_email-confirmation__button_9d1292"
    >
      Cancel
    </button>
  </div>
</DocumentFragment>
`;
